### DISCUSSION TITLE
How to use Zap in init()?

### DISCUSSION THREAD

**Discussion by albert-csms** [uber-go/zap#1115] [Q&A] [answered]
I have initialized Zap's SugaredLogger in `cmd/server.go`, and have been passing sugar to other files like so

```
	router.Handle(
		"/special_request",
		xray.Handler(
			xray.NewFixedSegmentNamer("my_service/special_request"),
			http.HandlerFunc(endpoint.SpecialRequestWrapper(sugar)),
		))
```

However, in one of my endpoints, I need to use zap in `init()`. I cannot use Go's `log` package because it does not provide logging at the `warning` level. How can zap be used in an `init()` function?

**Comment by abhinav** [comment, marked as answer]
Hello! To use Zap in init, you'll have to initialize Zap before that init runs—probably in a module that imports it.

Following that, you need to have the logger available globally. You can have your initialization logic put it in a global variable, or use Zap's global loggers.

To use Zap's global loggers, call [zap.ReplaceGlobals](https://pkg.go.dev/go.uber.org/zap#ReplaceGlobals) with your initialized logger. You can then access the SugaredLogger at [zap.S()](https://pkg.go.dev/go.uber.org/zap#S).

Closing because answered. Feel free to re-open or create a new issue if you have more questions. Thanks!
